pandas/core/reshape/merge.py

```diff
@@ -758,12 +758,26 @@ def _get_join_info(self):
             if self.right_index:
                 if len(self.left) > 0:
                     join_index = self.left.index.take(left_indexer)
+                    if (self.how == 'right' and -1 in left_indexer
+                            and not isinstance(self.right.index, MultiIndex)):
+                        join_list = join_index.to_numpy()
+                        absent = left_indexer == -1
+                        join_list[absent] = self.right.index.to_numpy()[absent]
+                        join_index = Index(join_list, dtype=join_index.dtype,
+                                           name=join_index.name)
```

Comment:
Let's add a quick comment as to what's going on here (and make sure to note that it applies symmetrically if it was the situation that you account for below).